pal/src/httpapi_sl.c

```diff
@@ -0,0 +1,492 @@
+            relativePath, (const char *)content, contentLength, 0);
+    if (ret < 0) {
+        LogError("HTTPClient_sendRequest failed, ret=%d", ret);
+        return (HTTPAPI_SEND_REQUEST_FAILED);
+    }
+
+    *statusCode = (unsigned int)ret;
+
+    /* Get the response headers */
+    cnt = 0;
+    offset = 0;
+
+    /*
+     * TODO: If there is more than one header of the same name, is this
+     * an issue with HTTPClient?
+     */
+    for (i = 0; i <= HTTPClient_MAX_RESPONSE_HEADER_FILEDS; i++) {
+        contentBufLen = CONTENT_BUF_LEN;
+        ret = HTTPClient_getHeader(cli, i, contentBuf, &contentBufLen, 0);
+        if (ret == HTTPClient_EGETOPTBUFSMALL) {
+            /* TODO: content buffer is too small. Enlarge and try again?? */
+            LogError("Content buffer is too small for incoming header");
+            ret = HTTPAPI_HTTP_HEADERS_FAILED;
+            goto headersDone;
+        }
+        else if (ret < 0) {
+            LogError("Failed to get header, ret=%d", ret);
+            ret = HTTPAPI_HTTP_HEADERS_FAILED;
+            goto headersDone;
+        }
+        else if (contentBufLen == 0) {
+            /* No data for this header */
+            continue;
+        }
+
+        ret = HTTPHeaders_AddHeaderNameValuePair(responseHeadersHandle,
+                HEADER_TO_STR(i), contentBuf);
+        if (ret != HTTP_HEADERS_OK) {
+            LogError("Adding the response header failed");
+            ret = HTTPAPI_HTTP_HEADERS_FAILED;
+            goto headersDone;
+        }
+        offset = 0;
+    }
+
+headersDone:
+    hname = NULL;
+    if (ret != 0) {
+        return ((HTTPAPI_RESULT)ret);
+    }
+
+    /* Get response body */
+    if (responseContent != NULL) {
+        offset = 0;
+        cnt = 0;
+
+        do {
+            ret = HTTPClient_readResponseBody(cli, contentBuf, CONTENT_BUF_LEN,
+                    &moreFlag);
+
+            if (ret < 0) {
+                LogError("HTTP read response body failed, ret=%d", ret);
+                ret = HTTPAPI_RECEIVE_RESPONSE_FAILED;
+                goto contentDone;
+            }
+
+            if (ret != 0) {
+                cnt = ret;
+                ret = BUFFER_enlarge(responseContent, cnt);
+                if (ret != 0) {
+                    LogError("Failed enlarging response buffer");
+                    ret = HTTPAPI_ALLOC_FAILED;
+                    goto contentDone;
+                }
+
+                ret = BUFFER_content(responseContent,
+                        (const unsigned char **)&hname);
+                if (ret != 0) {
+                    LogError("Failed getting the response buffer content");
+                    ret = HTTPAPI_ALLOC_FAILED;
+                    goto contentDone;
+                }
+
+                memcpy(hname + offset, contentBuf, cnt);
+                offset += cnt;
+            }
+        } while (moreFlag);
+
+    contentDone:
+        if (ret < 0) {
+            BUFFER_unbuild(responseContent);
```

Comment:
This doesn't catch all errors. `ret` will be < 0 if `HTTPClient_readResponseBody()` fails, but it will be != 0 if we failed to enlarge the buffer or get its content. (`HTTPAPI_ALLOC_FAILED` and `HTTPAPI_RECEIVE_RESPONSE_FAILED` are nonzero, but it they are _not_ negative.) At the same time, it's likely that the last call to `HTTPClient_readResponseBody()` will return a positive number (unless its API specifies otherwise). They may need to use separate variables, which would make sense anyway because the `HTTPClient_*` APIs return an `int` while our error codes are `HTTPAPI_RESULT`s. They already have a `cnt` variable they use anyway, which seems like it would be a natural choice.

Reply:
I agree. The ret variable has been dual purposed and makes the code less readable.

---
In reply to: [227941233](https://github.com/massand/review_pal/pull/1#discussion_r227941233) [](ancestors = 227941233)